In [2]:
import numpy as np
import pandas as pdx
import matplotlib.pyplot as plt

from pathlib import Path

In [3]:
dataset_path = Path("../dataset")
dl_data_path = dataset_path / "dl_data"

print("DL data path:")
print(dl_data_path)

DL data path:
..\dataset\dl_data


In [4]:
X_train = np.load(dl_data_path / "X_train.npy")
X_validation = np.load(dl_data_path / "X_validation.npy")
X_test = np.load(dl_data_path / "X_test.npy")

y_train = np.load(dl_data_path / "y_train.npy")
y_validation = np.load(dl_data_path / "y_validation.npy")
y_test = np.load(dl_data_path / "y_test.npy")

print("Datasets loaded successfully.")

Datasets loaded successfully.


In [5]:
print("===== INPUT SHAPES =====")

print("X_train:", X_train.shape)
print("X_validation:", X_validation.shape)
print("X_test:", X_test.shape)

print("\n===== LABEL SHAPES =====")

print("y_train:", y_train.shape)
print("y_validation:", y_validation.shape)
print("y_test:", y_test.shape)

===== INPUT SHAPES =====
X_train: (5969, 9000)
X_validation: (1279, 9000)
X_test: (1280, 9000)

===== LABEL SHAPES =====
y_train: (5969,)
y_validation: (1279,)
y_test: (1280,)


In [6]:
X_train = X_train.astype(np.float32)
X_validation = X_validation.astype(np.float32)
X_test = X_test.astype(np.float32)

print("Data type after conversion:")

print("X_train:", X_train.dtype)
print("X_validation:", X_validation.dtype)
print("X_test:", X_test.dtype)

Data type after conversion:
X_train: float32
X_validation: float32
X_test: float32


In [7]:
train_mean = X_train.mean()
train_std = X_train.std()

print("Training mean:", train_mean)
print("Training std:", train_std)

Training mean: 7.051846
Training std: 235.14249


In [8]:
X_train = (X_train - train_mean) / train_std
X_validation = (X_validation - train_mean) / train_std
X_test = (X_test - train_mean) / train_std

print("Normalization completed.")

Normalization completed.


In [9]:
print("===== NORMALIZATION CHECK =====")

print("Train:")
print("Mean:", X_train.mean())
print("Std:", X_train.std())

print("\nValidation:")
print("Mean:", X_validation.mean())
print("Std:", X_validation.std())

print("\nTest:")
print("Mean:", X_test.mean())
print("Std:", X_test.std())

===== NORMALIZATION CHECK =====
Train:
Mean: 8.628554e-08
Std: 1.0000006

Validation:
Mean: 0.00015843338
Std: 1.0341631

Test:
Mean: 0.0026376029
Std: 1.0162377


In [10]:
X_train = X_train[..., np.newaxis]
X_validation = X_validation[..., np.newaxis]
X_test = X_test[..., np.newaxis]

print("===== CNN INPUT SHAPES =====")

print("X_train:", X_train.shape)
print("X_validation:", X_validation.shape)
print("X_test:", X_test.shape)

===== CNN INPUT SHAPES =====
X_train: (5969, 9000, 1)
X_validation: (1279, 9000, 1)
X_test: (1280, 9000, 1)


In [11]:
import tensorflow as tf

print("TensorFlow version:", tf.__version__)
print("GPU available:", tf.config.list_physical_devices("GPU"))

TensorFlow version: 2.10.0
GPU available: [PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]


In [12]:
import sys

print("Python version:")
print(sys.version)

Python version:
3.10.10 (tags/v3.10.10:aad5f6a, Feb  7 2023, 17:20:36) [MSC v.1929 64 bit (AMD64)]


In [13]:
import subprocess

result = subprocess.run(
    ["nvidia-smi"],
    capture_output=True,
    text=True
)

print(result.stdout)
print(result.stderr)

Sun Aug 30 12:56:40 2026       
+-----------------------------------------------------------------------------------------+
| NVIDIA-SMI 555.97                 Driver Version: 555.97         CUDA Version: 12.5     |
|-----------------------------------------+------------------------+----------------------+
| GPU  Name                  Driver-Model | Bus-Id          Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |           Memory-Usage | GPU-Util  Compute M. |
|                                         |                        |               MIG M. |
|=========================================+========================+======================|
|   0  NVIDIA GeForce GTX 1650 ...  WDDM  |   00000000:01:00.0 Off |                  N/A |
| N/A   52C    P0             13W /   36W |       0MiB /   4096MiB |      0%      Default |
|                                         |                        |                  N/A |
+-----------------------------------------+-----

py -3.10 -m venv .venv_gpu
Set-ExecutionPolicy RemoteSigned -Scope CurrentUser
.venv_gpu\Scripts\activate
python -m pip install pandas matplotlib scikit-learn
python -m pip install "tensorflow==2.10.0"
python -m pip install "numpy<2"
python -m pip install ipykernel
python -m ipykernel install --user --name ecg-gpu --display-name "ECG GPU (TF 2.10)"


In [14]:
import sys
import tensorflow as tf

print("Python:")
print(sys.executable)

print("\nTensorFlow:")
print(tf.__version__)

print("\nGPU:")
print(tf.config.list_physical_devices("GPU"))

Python:
c:\Users\Anexpectant\Documents\aut\karshenasi\final_project\projects\main_project\Single_Lead_ECG_AF_Classification\.venv_gpu\Scripts\python.exe

TensorFlow:
2.10.0

GPU:
[PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]


******************************************************
START MOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOODEL
******************************************************

In [15]:
import tensorflow as tf
from tensorflow.keras import layers, models

print("TensorFlow:", tf.__version__)
print("GPU:", tf.config.list_physical_devices("GPU"))

TensorFlow: 2.10.0
GPU: [PhysicalDevice(name='/physical_device:GPU:0', device_type='GPU')]


In [16]:
model = models.Sequential([
    
    layers.Input(shape=(9000, 1)),
    
    layers.Conv1D(
        filters=32,
        kernel_size=7,
        activation="relu"
    ),
    layers.MaxPooling1D(pool_size=2),
    
    layers.Conv1D(
        filters=64,
        kernel_size=5,
        activation="relu"
    ),
    layers.MaxPooling1D(pool_size=2),
    
    layers.Conv1D(
        filters=128,
        kernel_size=3,
        activation="relu"
    ),
    layers.MaxPooling1D(pool_size=2),
    
    layers.GlobalAveragePooling1D(),
    
    layers.Dense(64, activation="relu"),
    
    layers.Dense(4, activation="softmax")
])

In [17]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv1d (Conv1D)             (None, 8994, 32)          256       
                                                                 
 max_pooling1d (MaxPooling1D  (None, 4497, 32)         0         
 )                                                               
                                                                 
 conv1d_1 (Conv1D)           (None, 4493, 64)          10304     
                                                                 
 max_pooling1d_1 (MaxPooling  (None, 2246, 64)         0         
 1D)                                                             
                                                                 
 conv1d_2 (Conv1D)           (None, 2244, 128)         24704     
                                                                 
 max_pooling1d_2 (MaxPooling  (None, 1122, 128)        0

In [18]:
model.compile(
    optimizer=tf.keras.optimizers.Adam(learning_rate=0.001),
    loss="sparse_categorical_crossentropy",
    metrics=["accuracy"]
)

print("Model compiled successfully.")

Model compiled successfully.


In [20]:
from sklearn.utils.class_weight import compute_class_weight
import numpy as np

classes = np.unique(y_train)

class_weights = compute_class_weight(
    class_weight="balanced",
    classes=classes,
    y=y_train
)

class_weight_dict = dict(zip(classes, class_weights))

print("===== CLASS WEIGHTS =====")

for class_id, weight in class_weight_dict.items():
    print(f"Class {class_id}: {weight:.4f}")

===== CLASS WEIGHTS =====
Class 0: 0.4200
Class 1: 2.8103
Class 2: 0.8830
Class 3: 7.6526


In [21]:
from tensorflow.keras.callbacks import EarlyStopping, ModelCheckpoint

early_stopping = EarlyStopping(
    monitor="val_loss",
    patience=8,
    restore_best_weights=True,
    verbose=1
)

model_checkpoint = ModelCheckpoint(
    "best_baseline_1d_cnn.keras",
    monitor="val_loss",
    save_best_only=True,
    verbose=1
)

In [22]:
history = model.fit(
    X_train,
    y_train,
    validation_data=(X_validation, y_validation),
    epochs=50,
    batch_size=32,
    class_weight=class_weight_dict,
    callbacks=[
        early_stopping,
        model_checkpoint
    ],
    verbose=1
)

Epoch 1/50
187/187 [==============================] - ETA: 0s - loss: 1.3291 - accuracy: 0.1404
Epoch 1: val_loss improved from inf to 1.34717, saving model to best_baseline_1d_cnn.keras
187/187 [==============================] - 40s 54ms/step - loss: 1.3291 - accuracy: 0.1404 - val_loss: 1.3472 - val_accuracy: 0.2901
Epoch 2/50
186/187 [============================>.] - ETA: 0s - loss: 1.2660 - accuracy: 0.2103
Epoch 2: val_loss improved from 1.34717 to 1.27103, saving model to best_baseline_1d_cnn.keras
187/187 [==============================] - 10s 53ms/step - loss: 1.2658 - accuracy: 0.2108 - val_loss: 1.2710 - val_accuracy: 0.2674
Epoch 3/50
186/187 [============================>.] - ETA: 0s - loss: 1.2286 - accuracy: 0.1857
Epoch 3: val_loss improved from 1.27103 to 1.25195, saving model to best_baseline_1d_cnn.keras
187/187 [==============================] - 9s 49ms/step - loss: 1.2280 - accuracy: 0.1856 - val_loss: 1.2519 - val_accuracy: 0.1048
Epoch 4/50
187/187 [=============